src/basic/exec-util.h

```diff
@@ -17,6 +17,23 @@
   along with systemd; If not, see <http://www.gnu.org/licenses/>.
 ***/
 
+#include <stdbool.h>
+
 #include "time-util.h"
 
-void execute_directories(const char* const* directories, usec_t timeout, char *argv[]);
+typedef int (*gather_stdout_callback_t) (int fd, void *arg);
+
+typedef struct {
+        gather_stdout_callback_t one;   /* from generators to helper process */
+        void *one_arg;
+        gather_stdout_callback_t two;   /* from helper process to main process */
+        void *two_arg;
+        gather_stdout_callback_t three; /* process data in main process */
+        void *three_arg;
+} gather_stdout_callbacks;
```

Comment:
Also, when I see something like this I wonder if this should better be a structure containing an array of callback/userdata pairs. After all you are even numerically indexing the entries by calling them "one", "two" and "three".

You could then define an enum with named indexes. Something like this:
```c
enum {
        STDOUT_GENERATE,
        STDOUT_COLLECT,
        STDOUT_CONSUME,
};
```

I think naming the three callbacks like this would make a lot of sense. Of course, the names aren't entirely self-explanatory, but I think "generate", "collect" and "consume" still are better than "one", "to", "three" ;-)